capa/features/extractors/dexfile.py

```diff
@@ -0,0 +1,372 @@
+                )
+            )
+
+        # Fill in the missing method data
+        for clazz in self.classes.values():
+            for method_def in clazz.data["direct_methods"]:
+                diff = method_def["diff"]
+                methods[diff].access_flags = method_def["access_flags"]
+                methods[diff].code_offset = method_def["code_off"]
+
+            for method_def in clazz.data["virtual_methods"]:
+                diff = method_def["diff"]
+                methods[diff].access_flags = method_def["access_flags"]
+                methods[diff].code_offset = method_def["code_off"]
+
+        # Fill in the missing code offsets with fake data
+        offset = self.dex.header_data["method_ids_off"]
+        for index, method in enumerate(methods):
+            if not method.has_code:
+                method.code_offset = offset + index * 8
+
+        return methods
+
+    def extract_file_features(self) -> Iterator[Tuple[Feature, Address]]:
+        yield Format(FORMAT_DEX), NO_ADDRESS
+
+        for method in self.methods:
+            if method.has_code:
+                yield FunctionName(method.qualified_name), DexMethodAddress(method.code_offset)
+            else:
+                yield FunctionName(method.qualified_name), NO_ADDRESS
+
+        for namespace in self.namespaces:
+            yield Namespace(namespace), NO_ADDRESS
+
+        for clazz in self.classes.values():
+            yield Class(clazz.class_type), DexClassAddress(clazz.offset)
+
+        for class_type in self.used_classes:
+            yield Class(class_type), NO_ADDRESS
+
+
+class DexFeatureExtractor(StaticFeatureExtractor):
+    def __init__(self, path: Path, *, code_analysis: bool):
+        super().__init__(hashes=SampleHashes.from_bytes(path.read_bytes()))
+        self.path: Path = path
+        self.code_analysis = code_analysis
+        self.dex = DEXParser(filedir=str(path))
+        self.analysis = DexAnalysis(self.dex)
+
+        # Perform more expensive code analysis only when requested
+        if self.code_analysis:
+            self.analysis.analyze_code()
+
+    def todo(self):
+        import inspect
+
+        message = "[DexparserFeatureExtractor:TODO] " + inspect.stack()[1].function
+        logger.debug(message)
+
+    def get_base_address(self):
+        return NO_ADDRESS
+
+    def extract_global_features(self) -> Iterator[Tuple[Feature, Address]]:
+        # These are hardcoded global features
+        yield Format(FORMAT_DEX), NO_ADDRESS
+        yield OS(OS_ANDROID), NO_ADDRESS
+        yield Arch(ARCH_DALVIK), NO_ADDRESS
+
+    def extract_file_features(self) -> Iterator[Tuple[Feature, Address]]:
+        yield from self.analysis.extract_file_features()
+
+    def is_library_function(self, addr: Address) -> bool:
+        assert isinstance(addr, DexMethodAddress)
+        method = self.analysis.methods_by_address[addr]
+        # exclude androidx/kotlin stuff?
+        return not method.has_code
```

Comment:
It looks like we are marking all methods as library functions? `capa` does not process methods that an extractor marks as a library function, see https://github.com/mandiant/capa/blob/d6f7d2180f47dbe3c3ae267374f84b36bb910a31/capa/capabilities/static.py#L165-L175

`is_library_function` is intended to detect statically linked library functions, e.g. `memcpy`, using whatever detection algorithm the corresponding extractor supports. It is not required for an extractor to support library detection. For example, our vivisect backend uses a `FLIRT` implementation while our `.NET` backend does not support library detection.

[`StaticFeatureExtractor.is_library_function`](https://github.com/mrexodia/capa/blob/a04ecb93fc904288804ec137dffb701d36d8fac5/capa/features/extractors/base_extractor.py#L175) returns `False` so there is no need to define the function here if library detection is not supported.

Reply:
I changed the `DexAnalyzedMethod` class to be more explicit about this. It now uses `return not method.has_definition` to filter out the methods that are not defined in the DEX file's classes. Additionally I added a `method.address` that always points to the most meaningful address (if there is a `code_offset` it points there, otherwise it points into the method_ids table). The idea is that you can follow the offset in IDA.